File: src/org/cappuccino_project/ide/intellij/plugin/indices/ObjJVariableNameByScopeIndex.java
package org.cappuccino_project.ide.intellij.plugin.indices;

import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.TextRange;
import com.intellij.psi.search.GlobalSearchScope;
import com.intellij.psi.stubs.StubIndexKey;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJBlock;
import org.cappuccino_project.ide.intellij.plugin.psi.ObjJVariableName;
import org.cappuccino_project.ide.intellij.plugin.utils.ObjJFileUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class ObjJVariableNameByScopeIndex extends ObjJStringStubIndexBase<ObjJVariableName> {
    private static final ObjJVariableNameByScopeIndex INSTANCE = new ObjJVariableNameByScopeIndex();
    public static final StubIndexKey<String, ObjJVariableName> KEY = IndexKeyUtil.createIndexKey(ObjJVariableNameByScopeIndex.class);
    private static final String KEY_FORMAT = "%s-%s-%s";
    private static final int VERSION = 1;


    private ObjJVariableNameByScopeIndex(){}

    public static ObjJVariableNameByScopeIndex getInstance() {
        return INSTANCE;
    }

    @Override
    public int getVersion() {
        return super.getVersion() + VERSION;
    }

    @NotNull
    @Override
    protected Class<ObjJVariableName> getIndexedElementClass() {
        return ObjJVariableName.class;
    }

    @NotNull
    @Override
    public StubIndexKey<String, ObjJVariableName> getKey() {
        return KEY;
    }

    @NotNull
    public List<ObjJVariableName> getInFile(@NotNull String fileName, @NotNull Project project) {
        return getInFile(fileName, null, project);
    }

    @NotNull
    public List<ObjJVariableName> getInFile(@NotNull String fileName, @Nullable GlobalSearchScope searchScope, @NotNull Project project) {
        return get(fileName+"-ALL", project, searchScope);
    }

    @NotNull
    public List<ObjJVariableName> getInRange(@NotNull String fileName, @NotNull TextRange textRange, @NotNull Project project) {
        return getInRange(fileName, textRange, null, project);
    }

    @NotNull
    public List<ObjJVariableName> getInRange(@NotNull String fileName, @NotNull TextRange textRange, @Nullable GlobalSearchScope searchScope, @NotNull Project project) {
        return getInRange(fileName, textRange.getStartOffset(), textRange.getEndOffset(), searchScope, project);
    }

    @NotNull
    public List<ObjJVariableName> getInRange(@NotNull String fileName, int elementStartOffset, int elementEndOffset, @NotNull Project project) {
        return getInRange(fileName, elementStartOffset, elementEndOffset, null, project);
    }

    @NotNull
    public List<ObjJVariableName> getInRange(@NotNull String fileName, int elementStartOffset, int elementEndOffset, @Nullable
            GlobalSearchScope scope, @NotNull Project project) {
        String queryKey = getIndexKey(fileName, elementStartOffset, elementEndOffset);
        return get(queryKey, project, scope);
        //return getAllForKeysFlat(getKeysInRange(fileName, elementStartOffset, elementEndOffset, project), project, scope);
    }

    @NotNull
    private List<String> getKeysInRange(@NotNull String fileName, int blockStart, int blockEnd, @NotNull Project project) {
        return getKeysInRange(fileName, blockStart, blockEnd, null, project);
    }

    @NotNull
    private List<String> getKeysInRange(@NotNull String fileName, int blockStart, int blockEnd, @Nullable GlobalSearchScope searchScope, @NotNull Project project) {
        List<String> keys = new ArrayList<>();
        String queryKey = getIndexKey(fileName, "(\\d+)", "(\\d+)");

        final Pattern pattern = Pattern.compile(queryKey);
        Matcher matcher;
        for (String key : getKeysByPattern(queryKey, project)) {
            matcher = pattern.matcher(key);
            if (matcher.groupCount() < 3) {
                continue;
            }
            int startOffset = Integer.parseInt(matcher.group(1));
            if (blockStart < startOffset) {
                continue;
            }
            int endOffset = Integer.parseInt(matcher.group(2));
            if (endOffset <= blockEnd) {
                keys.add(key);
            }
        }
        return keys;
    }

    @NotNull
    public static String getIndexKey(@NotNull
                                              ObjJBlock block) {
        final String fileName = ObjJFileUtil.getContainingFileName(block.getContainingFile());
        final TextRange textRange = block.getTextRange();
        final int startOffset = textRange.getStartOffset();
        final int endOffset = textRange.getEndOffset();
        return getIndexKey(fileName, startOffset,endOffset);
    }

    public static String getIndexKey(String fileName, Pair<Integer,Integer> blockRange) {
        return getIndexKey(fileName, blockRange.getFirst(), blockRange.getSecond());
    }

    public static String getIndexKey(String fileName, int startOffset, int endOffset) {
        return getIndexKey(fileName, startOffset+"", endOffset+"");
    }

    private static String getIndexKey(String fileName, String startOffset, String endOffset) {
        return String.format(KEY_FORMAT, fileName, startOffset, endOffset);
    }
}

File: src/cappuccino/ide/intellij/plugin/indices/ObjJVariableNameByScopeIndex.kt
package cappuccino.ide.intellij.plugin.indices

import cappuccino.ide.intellij.plugin.psi.ObjJVariableName
import cappuccino.ide.intellij.plugin.stubs.ObjJStubVersions
import com.intellij.openapi.project.Project
import com.intellij.openapi.util.Pair
import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiFile
import com.intellij.psi.search.GlobalSearchScope
import com.intellij.psi.stubs.AbstractStubIndex
import com.intellij.psi.stubs.StubIndexKey
import com.intellij.util.io.KeyDescriptor
import java.io.DataInput
import java.io.DataOutput

@Suppress("MemberVisibilityCanBePrivate")
class ObjJVariableNameByScopeIndex private constructor() : AbstractStubIndex<ObjJRangeKey, ObjJVariableName>() {

    override fun getVersion(): Int {
        return ObjJStubVersions.SOURCE_STUB_VERSION + VERSION
    }

    override fun getKey(): StubIndexKey<ObjJRangeKey, ObjJVariableName> {
        return KEY
    }

    fun getInRangeStrict(file: PsiFile, textRange: TextRange, project: Project): List<ObjJVariableName> {
        val fileScope = GlobalSearchScope.fileScope(file)
        return getAllKeys(project)
                .flatMap {
                    getInRangeStrict(it.variableName, textRange, fileScope, project)
                }
    }

    fun getInRangeStrict(file: PsiFile, variableName: String, textRange: TextRange, project: Project): List<ObjJVariableName> {
        val fileScope = GlobalSearchScope.fileScope(file)
        return getInRangeStrict(variableName, textRange, fileScope, project)
    }

    fun getInRangeStrict(variableName: String, textRange: TextRange, searchScope: GlobalSearchScope, project: Project): List<ObjJVariableName> {
        return getInRangeStrict(variableName, textRange.startOffset, textRange.endOffset, searchScope, project)
    }

    fun getInRangeStrict(variableName: String, elementStartOffset: Int, elementEndOffset: Int, scope: GlobalSearchScope, project: Project): List<ObjJVariableName> {
        val queryKey = ObjJRangeKey(variableName, elementStartOffset, elementEndOffset)
        return get(queryKey, project, scopeOrDefault(scope, project)).toList()
    }

    /*

    fun getInRangeFuzzy(file:PsiFile, textRange: TextRange, project: Project): List<ObjJVariableName> {
        return getInRangeFuzzy(file, textRange.startOffset, textRange.endOffset, project)
    }

    fun getInRangeFuzzy(file:PsiFile, startOffset: Int, endOffset:Int, project: Project): List<ObjJVariableName> {
        val fileScope = GlobalSearchScope.fileScope(file)
        return getAll(project, fileScope).filter {
            val variableTextRange = it.textRange
            variableTextRange.startOffset >= startOffset && variableTextRange.endOffset <= endOffset
        }
    }

    fun getInRangeFuzzy(variableName: String, textRange: TextRange, searchScope: GlobalSearchScope, project: Project): List<ObjJVariableName> {
        return getInRangeFuzzy(variableName, textRange.startOffset, textRange.endOffset, searchScope, project)
    }

    fun getInRangeFuzzy(variableName: String, elementStartOffset: Int, elementEndOffset: Int, scope: GlobalSearchScope, project: Project): List<ObjJVariableName> {
        val keys = filterKeys(variableName, elementStartOffset, elementEndOffset, project)
        val out = mutableListOf<ObjJVariableName>()
        for (key in keys) {
            out.addAll(get(key, project, scopeOrDefault(scope, project)))
        }
        return out
    }

    fun filterKeys(variableName: String, elementStartOffset: Int, elementEndOffset: Int, project: Project): List<ObjJRangeKey> {
        return getAllKeys(project).filter filter@{
            if (variableName != it.variableName)
                return@filter false
            if (elementStartOffset < it.start)
                return@filter false
            elementEndOffset <= it.end
        }
    }
     */

    companion object {
        val instance = ObjJVariableNameByScopeIndex()
        val KEY = IndexKeyUtil.createIndexKey(ObjJVariableNameByScopeIndex::class.java)
        private const val KEY_FORMAT = "%s:%s-%s"
        private const val VERSION = 1
        fun getIndexKey(fileName: String, blockRange: Pair<Int, Int>): String {
            return getIndexKey(fileName, blockRange.getFirst(), blockRange.getSecond())
        }

        fun getIndexKey(fileName: String?, startOffset: Int, endOffset: Int): String {
            return getIndexKey(fileName, startOffset.toString() + "", endOffset.toString() + "")
        }

        private fun getIndexKey(fileName: String?, startOffset: String, endOffset: String): String {
            return String.format(KEY_FORMAT, fileName, startOffset, endOffset)
        }

        internal fun scopeOrDefault(scope : GlobalSearchScope?, project: Project) : GlobalSearchScope {
            return scope ?: GlobalSearchScope.allScope(project)
        }
    }

    override fun getKeyDescriptor(): KeyDescriptor<ObjJRangeKey> {
        return object:KeyDescriptor<ObjJRangeKey> {
            override fun save(stream: DataOutput, key: ObjJRangeKey?) {
                stream.writeInt(key?.start ?: -1)
                stream.writeInt(key?.end ?: -1)
                stream.writeUTF(key?.variableName ?: "")
            }

            override fun isEqual(aKey: ObjJRangeKey?, otherKey: ObjJRangeKey?): Boolean {
                return aKey == otherKey
            }

            override fun getHashCode(key: ObjJRangeKey?): Int {
                return key.hashCode()
            }

            override fun read(stream: DataInput): ObjJRangeKey {
                val start = stream.readInt()
                val end = stream.readInt()
                val variableName = stream.readUTF()
                return ObjJRangeKey(variableName, start, end)
            }

        }
    }
}

data class ObjJRangeKey(val variableName:String, val start:Int, val end:Int)